sql/src/main/java/org/apache/druid/sql/calcite/planner/Rules.java

```diff
@@ -85,6 +88,13 @@
   public static final int DRUID_CONVENTION_RULES = 0;
   public static final int BINDABLE_CONVENTION_RULES = 1;
 
+  // Due to Calcite bug (CALCITE-3845), ReduceExpressionsRule can considered expression which is the same as the
+  // previous input expression as reduced. Basically, the expression is actually not reduced but is still considered as
+  // reduced. Hence, this resulted in an infinite loop of Calcite trying to reducing the same expression over and over.
+  // Calcite 1.23.0 fixes this issue by not consider expression as reduced if this case happens. However, while
+  // we are still using Calcite 1.21.0, a workaround is to limit the number of pattern matches to avoid infinite loop.
+  private static final int HEP_DEFAULT_MATCH_LIMIT = 1200;
```

Comment:
Is it possible to make this a system property? Having to restart the broker sounds like a lot less pain than having to make a new patch if this estimated `HEP_DEFAULT_MATCH_LIMIT` is incorrect for some edge case that we don't have tests for.

Reply:
Added a override system property `druid.sql.planner.hepMatchLimit` that can override the default value of 1200
This system property is not documented though as it should not be modify by user without deep investigation into Calcite/HepProgram use by Druid. Assuming you are doing investigation on determining new value for matchLimit in Druid code, then you will see this system property anyway.